Actions/BuildALDoc/BuildALDoc.HelperFunctions.ps1

```diff
@@ -0,0 +1,246 @@
+        }
+        else {
+            $ENV:aldocPath = Join-Path $tempFolder 'extension/bin/win32/aldoc.exe'
+        }
+
+        Write-Host "Installing/Updating docfx"
+        CmdDo -command dotnet -arguments @("tool","update","-g docfx")
+    }
+    $ENV:aldocPath
+}
+
+function SanitizeFileName([string] $filename) {
+    $filename.Replace('_','-').Replace('?','_').Replace('*','_').Replace(' ','-').Replace('\','-').Replace('/','-').Replace(':','-').Replace('<','-').Replace('>','-').Replace('|','-').Replace('%','pct')
+}
+
+function GetAppNameAndFolder {
+    Param(
+        [string] $appFile
+    )
+
+    $tmpFolder = Join-Path ([System.IO.Path]::GetTempPath()) ([Guid]::NewGuid().ToString())
+    Extract-AppFileToFolder -appFilename $appFile -appFolder $tmpFolder -generateAppJson
+    $appJson = Get-Content -Path (Join-Path $tmpFolder 'app.json') -Encoding utf8 | ConvertFrom-Json
+    $appJson.name
+    (SanitizeFileName -filename $appJson.name).ToLower()
+    Remove-Item -Path $tmpFolder -Recurse -Force
+}
+
+function GenerateDocsSite {
+    Param(
+        [string] $version,
+        [string[]] $allVersions,
+        [hashtable] $allApps,
+        [string] $repoName,
+        [string] $releaseNotes,
+        [string] $header,
+        [string] $footer,
+        [string] $defaultIndexMD,
+        [string] $defaultReleaseMD,
+        [string] $docsPath,
+        [string] $logLevel,
+        [switch] $hostIt
+    )
+
+    function ReplacePlaceHolders {
+        Param(
+            [string] $str,
+            [string] $version,
+            [string] $releaseNotes
+        )
+        return $str.Replace('{REPOSITORY}',$ENV:GITHUB_REPOSITORY).Replace('{VERSION}',$version).Replace('{RELEASENOTES}',$releaseNotes)
+    }
+
+    $indexTemplateRelativePath = '.aldoc/index.md'
+    if ($version) {
+        $thisTemplateRelativePath = '.aldoc/release.md'
+        $thisDefaultMD = $defaultReleaseMD
+    }
+    else {
+        $thisTemplateRelativePath = $indexTemplateRelativePath
+        $thisDefaultMD = $defaultIndexMD
+    }
+
+    $indexTemplatePath = Join-Path $ENV:GITHUB_WORKSPACE $thisTemplateRelativePath
+    if (-not (Test-Path $indexTemplatePath)) {
+        $indexTemplatePath = Join-Path $ENV:GITHUB_WORKSPACE $indexTemplateRelativePath
+    }
+    if (Test-Path $indexTemplatePath) {
+        $indexTemplate = Get-Content -Encoding utf8 -Path $indexTemplatePath -Raw
+    }
+    else {
+        $indexTemplate = $thisDefaultMD
+    }
+    $indexContent = ReplacePlaceHolders -str $indexTemplate.Replace('{INDEXTEMPLATERELATIVEPATH}',$thisTemplateRelativePath) -version $version -releaseNotes $releaseNotes
+
+    $alDocPath = DownloadAlDoc
+    $docfxPath = Join-Path ([System.IO.Path]::GetTempPath()) ([Guid]::NewGuid().ToString())
+    New-Item -path $docfxPath -ItemType Directory | Out-Null
+    try {
+        $apps = @()
+        # Generate new toc.yml and calculate apps - releases and projects
```

Comment:
Consider refactoring the generation of the table of content into a separate function :)